### ISSUE TITLE
`import/no-extraneous-dependencies` regression from 2.22.1 to 2.23.0 on Yarn 2 with npm dependency protocol

### ISSUE THREAD

**Issue by JLHwung** [benmosher/eslint-plugin-import#2072]
When I bump `eslint-plugin-import` from 2.22.0 to 2.23.0, the linter [bails](https://github.com/JLHwung/babel/runs/2587157601?check_suite_focus=true#step:8:91) under certain Yarn 2 `npm:` protocol dependencies

`package.json`:

```jsonc
{
  "devDependencies": {
    "@babel-baseline/parser": "npm:@babel/parser@^7.14.0" // "npm:" is a yarn 2 protocol
  }
}
```

Input code:
```js
import baseline from "@babel-baseline/parser";
// eslint complains
// '@babel/parser' should be listed in the project's dependencies. Run 'npm i -S @babel/parser' to add it (import/no-extraneous-dependencies)
```

ESLint config: https://github.com/JLHwung/babel/blob/c18d774916c7fdec04c5ef5b4cdc3be31cddc55f/.eslintrc.cjs

Although this issue might be specific to yarn 2, one can replace `npm:@babel/parser` to any local path supported also by npm, as long as the `name` in `package.json` (`@babel/parser`) does not equal to the value of import specifier (`@babel-baseline/parser`).

The linter is happy on 2.22.0 and 2.22.1.

Reproduction: https://github.com/JLHwung/babel/commit/c18d774916c7fdec04c5ef5b4cdc3be31cddc55f, clone this repo and run `yarn; make build; make lint`.

Might be related: https://github.com/benmosher/eslint-plugin-import/issues/1712

**Comment by ljharb** [comment]
`npm:` is also an npm protocol (the far more important consideration), so it's definitely something we should support.

However, I don't think it's necessarily the protocol (we don't look at the RHS), i think this might be the same issue as #2065 and #2066.

**Comment by paztis** [comment]
Here we've resolve you real dependency.
And it's real name is @babel/parser, not @babel-baseline/parser

Why do you change its name ?

**Comment by paztis** [comment]
You might use the module eslint-import-resolver-alias to explain the alias to @babel/parser is @babel-baseline/parser

**Comment by ljharb** [comment]
I don't think we should need a resolver alias here - this is a first-class feature of npm, and the node resolver itself should be supporting it.

**Comment by JLHwung** [comment]
> And it's real name is @babel/parser, not @babel-baseline/parser
> Why do you change its name ?

Good question! The import is renamed to make it clearly that we are benchmarking against the npm published version of `@babel/parser`, as a baseline, versus the current `@babel/parser` (resolved to `./packages/babel-parser/lib/index.js` in a yarn 2 workspace).

I agree with @ljharb, it would be great if I don't have to duplicate the alias info in both `package.json` and `.eslintrc`.

**Comment by paztis** [comment]
What is not correct in your solution is the name inside the package.json didn't fit to the name of the imports. 
Their real resolution has been done exactly to resolve the aliases (webpack, ...) That transform the name of the import.

It was the only correct way to resolve the real name of an import.

For your cas I don't know how we can resolve it correctly.
All the old process was bad because it didn't resolve correctly module names totally aliases, mostly in monorepo

**Comment by paztis** [comment]
after a little search, in your case we might need an enhancement in `src/rules/no-extraneous-dependencies.js` file

```
  // get the real name from the resolved package.json
  // if not aliased imports (alias/react for example) will not be correctly interpreted
  // fallback on original name in case no package.json found
  const packageName = getModuleRealName(resolved) || getModuleOriginalName(name);
```

Here we take the real name  from package.json (which is not the correct one at your level) and in fallback the import name.
I was thinking the `resolved` variable can be computed without resolving symlinks, but as the resolving is done in extra modules, we cannot have this guarantee.

It means we might check both here: real name `getModuleRealName(resolved)` and import name 'getModuleOriginalName(name)'.
If one of both is present in dependencies, it might be ok.

@benmosher , @ljharb , what do you think of this workaround ? It might resolve the problem of multiple people since the 2.23 delivery caused by package.json not in phase with the module name, bad extra package.json in nested folders, ...

**Comment by ljharb** [comment]
It's perfectly normal, and becoming more common, for nested folders to have package.json files - so any solution that doesn't account for that isn't going to be ideal.

**Comment by paztis** [comment]
So you want to do it ?

**Comment by ljharb** [comment]
If it will address the problem without introducing brittleness, definitely, and it'd be quite appreciated!

It's hard for me to know if it's a good solutiong or not prior to a PR, though.

**Comment by paztis** [comment]
Originally, this change was done because the algorithm to compute the package name from the imports was not resolving all the aliases then give you a bad dependency name.
So the changed was based on the resolved path provided by the resolvers.

Now if you force the resolvers to give you a non symlinked resolved path (it is possible with require.reaolve or FS to avoid this resolution), you can do an algorithm with node_module detection.

Now if not possible, all the user of monorepo are blocked.

Perhaps we can do something like: 
```
If (path.indexOf('node_modules') {...} else {...}
```

**Comment by ljharb** [comment]
We can use the `resolve` package but disable symlink resolution, if that's desired.

Certainly branching on "whether it's inside node_modules" seems reasonable.

**Comment by paztis** [comment]
this defect might be closed in 2.23.4, isn't it ?

**Comment by ljharb** [comment]
@JLHwung can you confirm?

**Comment by JLHwung** [comment]
Can confirm it is fixed in 2.23.4. Thanks!
